snapcraft/models/project.py

```diff
@@ -48,20 +48,21 @@
 ProjectName = Annotated[str, StringConstraints(max_length=40)]
 
 
-def _validate_command_chain(command_chains: list[str] | None) -> list[str] | None:
+def _validate_command_chain(command_chains: list[str]) -> list[str]:
     """Validate command_chain."""
-    if command_chains is not None:
-        for command_chain in command_chains:
-            if not re.match(r"^[A-Za-z0-9/._#:$-]*$", command_chain):
-                raise ValueError(
-                    f"{command_chain!r} is not a valid command chain. Command chain entries must "
-                    "be strings, and can only use ASCII alphanumeric characters and the following "
-                    "special characters: / . _ # : $ -"
-                )
+    for command_chain in command_chains:
+        if not re.match(r"^[A-Za-z0-9/._#:$-]*$", command_chain):
+            raise ValueError(
+                f"{command_chain!r} is not a valid command chain. Command chain entries must "
+                "be strings, and can only use ASCII alphanumeric characters and the following "
+                "special characters: / . _ # : $ -"
+            )
     return command_chains
 
 
-def validate_architectures(architectures):
+def validate_architectures(
```

Comment:
The functions related to architectures had a lot of type-related issues and had to get some refactoring.

I tried to strike a balance between refactoring and ignoring type checks.  There should be no behavioral changes.